articles/app-service/app-service-web-tutorial-connect-msi.md

```diff
@@ -136,29 +136,60 @@ Type `Ctrl+F5` to run the app again. The same CRUD app in your browser is now co
+{
+    private static readonly string[] _azureSqlScopes = new[]
+    {
+        "https://database.windows.net//.default"
+    };
+
+    private static readonly TokenCredential _credential = new DefaultAzureCredential();
+
+    public override Task<SqlAuthenticationToken> AcquireTokenAsync(SqlAuthenticationParameters parameters)
+    {
+        var tokenRequestContext = new TokenRequestContext(_azureSqlScopes);
+        var tokenResult = _credential.GetToken(tokenRequestContext, default);
+        return Task.FromResult(new SqlAuthenticationToken(tokenResult.Token, tokenResult.ExpiresOn));
+    }
+
+    public override bool IsSupported(SqlAuthenticationMethod authenticationMethod) => authenticationMethod.Equals(SqlAuthenticationMethod.ActiveDirectoryDeviceCodeFlow);
+}
+```
+In `Startup.cs` update the `ConfigureServices()` method with the following code:
+
+```csharp
+services.AddControllersWithViews();
+services.AddDbContext<MyDatabaseContext>(options =>
+{
+    SqlAuthenticationProvider.SetProvider(
+        SqlAuthenticationMethod.ActiveDirectoryDeviceCodeFlow, 
+        new CustomAzureSQLAuthProvider());
+    var sqlConnection = new SqlConnection(Configuration.GetConnectionString("MyDatabaseContext"));
+    options.UseSqlServer(sqlConnection);
+});
 ```
 
 > [!NOTE]
 > This demonstration code is synchronous for clarity and simplicity.
 
-That's every thing you need to connect to SQL Database. When debugging in Visual Studio, your code uses the Azure AD user you configured in [Set up Visual Studio](#set-up-visual-studio). You'll set up SQL Database later to allow connection from the managed identity of your App Service app. The `AzureServiceTokenProvider` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.
+The above code makes use of the `Azure.Identity` library so that it can authenticate and retrieve an access token for the database, no matter
+where the code is running. If you're running on your local machine, the `DefaultAzureCredential()` will loop through a number of options to 
+find a valid, logged in account. You can read more about the `DefaultAzureCredential` class [in the docs](https://docs.microsoft.com/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet)
 
-> [!TIP]
-> If the Azure AD user you configured has access to multiple tenants, call `GetAccessTokenAsync("https://database.windows.net/", tenantid)` with the desired tenant ID to retrieve the proper access token.
+That's every thing you need to connect to SQL Database. When debugging in Visual Studio, your code uses the Azure AD user you configured in [Set up Visual Studio](#set-up-visual-studio). You'll set up SQL Database later to allow connection from the managed identity of your App Service app. The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.
```

Comment:
```suggestion
That's everything you need to connect to SQL Database. When debugging in Visual Studio, your code uses the Azure AD user you configured in [Set up Visual Studio](#set-up-visual-studio). You'll set up SQL Database later to allow connection from the managed identity of your App Service app. The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.
```

> The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.

I'm fairly sure that's not the case, unless something changed recently.
My understanding is that `Azure.Identity` itself does not handle token caching / renewal; the Azure client libraries like Blobs, Queues, ServiceBus, etc... do this through a custom HTTP pipeline.

Happy to be proven wrong, though 😆 &mdash; @schaabs?

Reply:
> > The `DefaultAzureCredential` class caches the token in memory and retrieves it from Azure AD just before expiration. You don't need any custom code to refresh the token.
> 
> I'm fairly sure that's not the case, unless something changed recently.
> My understanding is that `Azure.Identity` itself does not handle token caching / renewal; the Azure client libraries like Blobs, Queues, ServiceBus, etc... do this through a custom HTTP pipeline.
> 
> Happy to be proven wrong, though 😆 — @schaabs?

In that case, I question the appropriateness of the change at all. The AppAuthentication approach so far is actually demonstrably simpler, especially since previously it was just 2 lines of c# code with no need for custom providers, AND it caches tokens. Now we're telling the customers now they need to write more code AND they need to add code to cache tokens? I don't mean to be harsh but this isn't an attractive scenario.

Reply:
I agree that having the user code have to cache the token is less than ideal, but not all credential types in Azure.Identity cache tokens. We hope to provide a better story around SQL authentication in the long term, most likely by providing a `SqlAuthenticationProvider` similar to what @cmatskas has provided in this sample, which would handle any caching for the user, but while that is on our roadmap we currently don't have the work planned.

That said I don't think you can compare this new sample using Azure.Identity, with the previous sample using the AppAuthentication library 1:1. The old sample simply acquired a token and statically set that token on the connection. So it would seem to me that the old sample wouldn't handle token expiration at all. But perhaps I'm wrong here, as I'm not totally familiar with how EF creates creates and reuses connections.

Reply:
@schaabs Sorry I had a long break in April. Can we compromise by going ahead with the change on my side but putting `SqlAuthenticationProvider` on the roadmap on your end?

I agree about not making 1:1 comparison with AppAuthentication. However I did read docs somewhere that said that AppAuthentication library handles caching by itself so the user doesn't have to. I don't think it's out of reason to expect the new tech to match that.

Reply:
Thanks for agreeing to unblock this @schaabs. I know that the team is currently working on a better SQLClient + Azure.identity experience, so good news on this front